pkg/metrics/reconcile_time_metrics.go

```diff
@@ -0,0 +1,90 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sync"
+	"time"
+)
+
+type ReconcileTimeMetrics struct {
+	reconcileTimeSeconds         *prometheus.GaugeVec
+	reconcileDeployTimeSeconds   *prometheus.GaugeVec
+	reconcileFetchTimeSeconds    *prometheus.GaugeVec
+	reconcileTemplateTimeSeconds *prometheus.GaugeVec
+}
+
+var (
+	timeMetricsOnce sync.Once
+)
+
+func NewReconcileTimeMetrics() *ReconcileTimeMetrics {
+	const (
+		metricNamespace     = "kappctrl_reconcile_time_seconds"
```

Comment:
can you please share the output of metrics endpoint? I think namespace should be just "kappctrl" otherwise it will duplicate and results in names like "kappctrl_reconcile_time_seconds_reconcile_fetch_time_seconds".

Reply:
```
# TYPE kappctrl_reconcile_deploy_time_seconds gauge
kappctrl_reconcile_deploy_time_seconds{controller="app",firstReconcile="false",name="pkg-demo",namespace="default"} 3.27706651
kappctrl_reconcile_deploy_time_seconds{controller="app",firstReconcile="false",name="simple-app",namespace="default"} 0.082659933
kappctrl_reconcile_deploy_time_seconds{controller="app",firstReconcile="true",name="simple-app",namespace="default"} 9.367542779
kappctrl_reconcile_deploy_time_seconds{controller="pkgr",firstReconcile="false",name="e2e-repo.test.carvel.dev",namespace="kapp-controller-packaging-global"} 0.375847592
kappctrl_reconcile_deploy_time_seconds{controller="pkgr",firstReconcile="true",name="e2e-repo.test.carvel.dev",namespace="kapp-controller-packaging-global"} 0.350623984
kappctrl_reconcile_deploy_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-100.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 2.324662531
kappctrl_reconcile_deploy_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-50.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.683961845
kappctrl_reconcile_deploy_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-500.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 15.71387306
kappctrl_reconcile_deploy_time_seconds{controller="pkgr",firstReconcile="true",name="probably-tap-or-something.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.17900214
# HELP kappctrl_reconcile_fetch_time_seconds Time taken to perform a fetch for a CR
# TYPE kappctrl_reconcile_fetch_time_seconds gauge
kappctrl_reconcile_fetch_time_seconds{controller="app",firstReconcile="false",name="pkg-demo",namespace="default"} 4.728198131
kappctrl_reconcile_fetch_time_seconds{controller="app",firstReconcile="false",name="pkg-demo",namespace="test"} 4.686686265
kappctrl_reconcile_fetch_time_seconds{controller="app",firstReconcile="false",name="simple-app",namespace="default"} 1.304734138
kappctrl_reconcile_fetch_time_seconds{controller="app",firstReconcile="true",name="pkg-demo",namespace="default"} 4.912922364
kappctrl_reconcile_fetch_time_seconds{controller="app",firstReconcile="true",name="pkg-demo",namespace="test"} 4.8636598620000004
kappctrl_reconcile_fetch_time_seconds{controller="app",firstReconcile="true",name="simple-app",namespace="default"} 1.299503442
kappctrl_reconcile_fetch_time_seconds{controller="pkgr",firstReconcile="false",name="e2e-repo.test.carvel.dev",namespace="kapp-controller-packaging-global"} 4.561660644
kappctrl_reconcile_fetch_time_seconds{controller="pkgr",firstReconcile="true",name="e2e-repo.test.carvel.dev",namespace="kapp-controller-packaging-global"} 5.551657015
kappctrl_reconcile_fetch_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-100.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.015032984
kappctrl_reconcile_fetch_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-50.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.013356157
kappctrl_reconcile_fetch_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-500.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.064673964
kappctrl_reconcile_fetch_time_seconds{controller="pkgr",firstReconcile="true",name="probably-tap-or-something.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.015370906
# HELP kappctrl_reconcile_template_time_seconds Time taken to perform a templating for a CR
# TYPE kappctrl_reconcile_template_time_seconds gauge
kappctrl_reconcile_template_time_seconds{controller="app",firstReconcile="false",name="pkg-demo",namespace="default"} 0.046433794
kappctrl_reconcile_template_time_seconds{controller="app",firstReconcile="false",name="pkg-demo",namespace="test"} 0.043730396
kappctrl_reconcile_template_time_seconds{controller="app",firstReconcile="false",name="simple-app",namespace="default"} 0.006354081
kappctrl_reconcile_template_time_seconds{controller="app",firstReconcile="true",name="pkg-demo",namespace="default"} 0.043066916
kappctrl_reconcile_template_time_seconds{controller="app",firstReconcile="true",name="pkg-demo",namespace="test"} 0.045291503
kappctrl_reconcile_template_time_seconds{controller="app",firstReconcile="true",name="simple-app",namespace="default"} 0.01640674
kappctrl_reconcile_template_time_seconds{controller="pkgr",firstReconcile="false",name="e2e-repo.test.carvel.dev",namespace="kapp-controller-packaging-global"} 0.278722011
kappctrl_reconcile_template_time_seconds{controller="pkgr",firstReconcile="true",name="e2e-repo.test.carvel.dev",namespace="kapp-controller-packaging-global"} 0.515563694
kappctrl_reconcile_template_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-100.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.114663111
kappctrl_reconcile_template_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-50.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.075671007
kappctrl_reconcile_template_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-500.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.594071466
kappctrl_reconcile_template_time_seconds{controller="pkgr",firstReconcile="true",name="probably-tap-or-something.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.038623431
# HELP kappctrl_reconcile_time_seconds Overall time taken to reconcile a CR
# TYPE kappctrl_reconcile_time_seconds gauge
kappctrl_reconcile_time_seconds{controller="app",firstReconcile="false",name="pkg-demo",namespace="default"} 8.065660199
kappctrl_reconcile_time_seconds{controller="app",firstReconcile="false",name="pkg-demo",namespace="test"} 4.73399651
kappctrl_reconcile_time_seconds{controller="app",firstReconcile="false",name="simple-app",namespace="default"} 1.394007851
kappctrl_reconcile_time_seconds{controller="app",firstReconcile="true",name="pkg-demo",namespace="default"} 5.224242495
kappctrl_reconcile_time_seconds{controller="app",firstReconcile="true",name="pkg-demo",namespace="test"} 4.98715186
kappctrl_reconcile_time_seconds{controller="app",firstReconcile="true",name="simple-app",namespace="default"} 10.712085359
kappctrl_reconcile_time_seconds{controller="pkgi",firstReconcile="false",name="pkg-demo",namespace="default"} 0.07437761
kappctrl_reconcile_time_seconds{controller="pkgi",firstReconcile="false",name="pkg-demo",namespace="test"} 0.018053127
kappctrl_reconcile_time_seconds{controller="pkgi",firstReconcile="true",name="pkg-demo",namespace="default"} 0.036914327
kappctrl_reconcile_time_seconds{controller="pkgi",firstReconcile="true",name="pkg-demo",namespace="test"} 0.041035891
kappctrl_reconcile_time_seconds{controller="pkgr",firstReconcile="false",name="e2e-repo.test.carvel.dev",namespace="kapp-controller-packaging-global"} 5.217053404
kappctrl_reconcile_time_seconds{controller="pkgr",firstReconcile="true",name="e2e-repo.test.carvel.dev",namespace="kapp-controller-packaging-global"} 6.4527741800000005
kappctrl_reconcile_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-100.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 3.003516656
kappctrl_reconcile_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-50.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 1.3014173549999999
kappctrl_reconcile_time_seconds{controller="pkgr",firstReconcile="true",name="minimal-repo-500.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 16.767015768
kappctrl_reconcile_time_seconds{controller="pkgr",firstReconcile="true",name="probably-tap-or-something.tanzu.carvel.dev",namespace="kapp-controller-packaging-global"} 0.317999077
```